atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java

```diff
@@ -169,6 +169,19 @@ boolean allowHiddenTableAccess() {
 
     abstract TaggedMetricRegistry globalTaggedMetricRegistry();
 
+    /**
+     * If config().initializeAsync() is set to true, the callback Runnable will be run when the TransactionManager is
+     * successfully initialized. The TransactionManager stay uninitialized and continue to throw for all other purposes
+     * until the callback returns at which point it will become initialized.
+     *
+     * Note that if the callback throws an exception or blocks forever, the TransactionManager will never become
```

Comment:
This can be read as "if the callback throws an exception the TransactionManager will never become initialized." which isn't what we want.

Does the callback need to catch its own failures and rollback any changes it made?

Reply:
I don't think we can safely catch and ignore exceptions thrown by the callback - I think it's down to the caller to (a) handle exceptions, and (b) decide whether to abort, retry (with or without rollback), or ignore.

Reply:
I think we need some way to get signal out. One use case for this is to check that the DB state agrees with what the client expects. If the schema doesn't match (example check in the callback), we'd like the service to be able to report as unhealthy and know that the initialization failed.

Reply:
@tboam Is it not what we want? If we do not want to say we are initialized until the callback is done, then we should fail on an exception. We cannot guess the caller's intention.

@sandorw We could store the exception from the callback so we can then log it, wrap it in the NotInitialized exception on method calls, or something similar.